### PR TITLE
istio upgrade e2e tests using either helm or istioctl

### PR DESCRIPTION
**Pull request by carolynhu**
https://github.com/istio/istio/issues/17649
https://github.com/istio/istio/issues/17670
This PR is relying on this new `istioctl x upgrade` command: https://github.com/istio/operator/pull/348

### CODE REVIEW

**Comment by elfinhe** [comment]
please add some output sample here.

File: tests/upgrade/test_crossgrade.sh [hunk 1]
```diff
@@ -227,9 +234,38 @@ installIstioSystemAtVersionHelmTemplate() {
...
+  fi
+}
+
+istioUpgradeOptions(){
+  if [[ "${INSTALL_OPTIONS}" == "helm" ]];then
+    installIstioAtVersionUsingHelm "${TO_HUB}" "${TO_TAG}" "${TO_PATH}"
+  else
+    upgradeIstioAtVersionUsingIstioctl "${TO_HUB}" "${TO_TAG}" "${TO_PATH}"
+  fi
 }
```
Comment (elfinhe) [inline tests/upgrade/test_crossgrade.sh]:
```bash
istioUpgradeOptions(){
  if [[ "${INSTALL_OPTIONS}" == "helm" ]];then
    installIstioAtVersionUsingHelm "${TO_HUB}" "${TO_TAG}" "${TO_PATH}"
  else if [[ "${INSTALL_OPTIONS}" == "istioctl" ]];then
    upgradeIstioAtVersionUsingIstioctl "${TO_HUB}" "${TO_TAG}" "${TO_PATH}"
  else
    exit 1
  fi
}
```

Comment (elfinhe) [reply tests/upgrade/test_crossgrade.sh]: if not an option, should we exit?

Comment (carolynhu) [reply tests/upgrade/test_crossgrade.sh]: good catch, we should only allow two cases here.

File: tests/upgrade/test_crossgrade.sh [hunk 2]
```diff
@@ -227,9 +234,44 @@ installIstioSystemAtVersionHelmTemplate() {
     --set-string global.tag="${2}" \
     --set global.defaultPodDisruptionBudget.enabled=true > "${ISTIO_ROOT}/istio.yaml" || die "helm template failed"
 
+    withRetries 3 60 kubectl apply -f "${ISTIO_ROOT}"/istio.yaml
```
Comment (howardjohn) [inline tests/upgrade/test_crossgrade.sh]: Why do we need to retry?

File: tests/upgrade/test_crossgrade.sh [hunk 3]
```diff
@@ -227,9 +234,44 @@ installIstioSystemAtVersionHelmTemplate() {
...
+}
 
-    withRetries 3 60 kubectl apply -f "${ISTIO_ROOT}"/istio.yaml
+# istioctl x upgrade supports upgrade istio release version
+# from 1.3.x to 1.3.y
+# from 1.3.x to 1.4.0
+upgradeIstioAtVersionUsingIstioctl(){
+  writeMsg "istioctl upgrade istio using version ${2} from ${3}."
+  istioctl_path="${3}"/bin
+  withRetries 3 60 "${istioctl_path}"/istioctl experimental manifest upgrade --skip-confirmation
```
Comment (howardjohn) [inline tests/upgrade/test_crossgrade.sh]: Why retry?

File: tests/upgrade/test_crossgrade.sh [hunk 4]
```diff
@@ -430,12 +472,12 @@ checkEchosrv
 
 # Run internal traffic in the background since we may have to relaunch it if the job fails.
 sendInternalRequestTraffic &
-sendExternalRequestTraffic "${INGRESS_ADDR}"
+sendExternalRequestTraffic "${INGRESS_ADDR}" &   # TODO: if we wait this to finish, all the following steps will suceed.
```
Comment (howardjohn) [inline tests/upgrade/test_crossgrade.sh]: if we wait then this is not really testing much, The whole point is that the traffic is sent during the update

**Comment by howardjohn** [comment]
/retest

Known flake https://github.com/istio/istio/issues/18385

File: tests/upgrade/test_crossgrade.sh (hunk 4 above)
Comment (carolynhu) [reply tests/upgrade/test_crossgrade.sh]: yes, that's why I have the `&` here to make the external traffic send during both installation and upgrade in background. Here is the issue related to this script. https://github.com/istio/istio/issues/18369

**Pull request merged**